FAST-LIO/src/GNSS_Processing.hpp

```diff
@@ -0,0 +1,129 @@
+#include <cmath>
+#include <math.h>
+#include <deque>
+#include <mutex>
+#include <thread>
+#include <fstream>
+#include <csignal>
+#include <ros/ros.h>
+#include <so3_math.h>
+#include <Eigen/Eigen>
+#include <common_lib.h>
+#include <pcl/common/io.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+#include <condition_variable>
+#include <nav_msgs/Odometry.h>
+#include <pcl/common/transforms.h>
+#include <pcl/kdtree/kdtree_flann.h>
+#include <tf/transform_broadcaster.h>
+#include <eigen_conversions/eigen_msg.h>
+#include <pcl_conversions/pcl_conversions.h>
+#include <sensor_msgs/Imu.h>
+#include <sensor_msgs/PointCloud2.h>
+#include <geometry_msgs/Vector3.h>
+#include "use-ikfom.hpp"
+
+#include <GeographicLib/LocalCartesian.hpp>                 //  Call the GeographicLib library
+
+class GnssProcess
+{
+ public:
+    EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+    double time ;
+    double latitude ;
+    double longitude ;
+    double altitude ;
+    double local_E ;
+    double local_N ;
+    double local_U ;
+    int status ;
+    int service ;
+
+    double origin_longitude;
+    double origin_latitude;
+    double origin_altitude;
+
+    V3D pose_cov ;
+
+    GnssProcess();
+    ~GnssProcess();
+
+    void InitOriginPosition(double latitude, double longitude, double altitude);
+    void UpdateXYZ(double latitude, double longitude, double altitude);
+
+    void Reverse(
+      const double &local_E, const double &local_N, const double &local_U,
+      double &lat, double &lon, double &alt
+    );
+
+    void set_extrinsic(const V3D &transl, const M3D &rot);
+    void set_extrinsic(const V3D &transl);
+    void set_extrinsic(const MD(4,4) &T);
+
+  private:
+    GeographicLib::LocalCartesian geo_converter;
+
+    M3D Gnss_R_wrt_Lidar ;
+    V3D Gnss_T_wrt_Lidar;
+};
+
+GnssProcess::GnssProcess()
+{
+    time = 0.0;
+    local_E = 0.0;
+    local_N = 0.0;
+    local_U = 0.0;
+    status = 0;
+    service = 0;
+
+    origin_longitude = 0 ;
+    origin_latitude = 0;
+    origin_altitude = 0;
+    pose_cov = Zero3d;
+    Gnss_T_wrt_Lidar = Zero3d;  
+    Gnss_R_wrt_Lidar = Eye3d;
+}
+
+GnssProcess::~GnssProcess() {}
+
+//! Initialize the origin, WGS84 -> ENU ??? The debugging result seems to be NED North East
```

Comment:
This just occurred to me, but you've been testing with a level lidar, whereas we fly with a tilted lidar. With tilted lidar, the slam_map frame is also tilted with respect to the earth's surface, so you'll need to account for that transform. However in the interest of time, let's leave that for after the ROS2 conversion. For now please add a launch arg specifying whether or not to use gps in slam :blush:

Reply:
Oh no! This completely slipped of my mind. Added it as a TODO!